File: scripts/samples/c/treenode_sol.c

#include "../inc/global.h"
#include "../inc/test.h"
#include "../utils/compare.h"
#include "../utils/print.h"

// Debug mode
#define DEBUG
#include "../inc/debug.h"

// Display info of the approach
#define INFO

// Approach
#define APR 1
#ifdef INFO
#if APR == 1
static char apr_idea[40] = "";
static char time_cmplx[20] = "";
static char space_cmplx[20] = "";
#endif
#endif

/*
Problem 1339: Maximum Product of Splitted Binary Tree
LeetCode: https://leetcode.com/problems/maximum-product-of-splitted-binary-tree/
*/
#if APR == 1
// Code
long long total_sum = 0;
long long max_product = 0;
const int MOD = 1e9 + 7;

long long _countTotalSum(struct TreeNode* node) {
    if (!node) return 0;
    return node->val + _countTotalSum(node->left) + _countTotalSum(node->right);
}

long long _countMaxProduct(struct TreeNode* node) {
    if (!node) return 0;

    long long left = _countMaxProduct(node->left);
    long long right = _countMaxProduct(node->right);
    long long sub = node->val + left + right;
    long long product = sub * (total_sum - sub);

    if (product > max_product) max_product = product;

    return sub;
}

int maxProduct(struct TreeNode* root) {
    max_product = 0;
    total_sum = _countTotalSum(root);
    _countMaxProduct(root);
    return max_product % MOD;
}
#endif

void test1339() {
    #ifdef INFO
    printf("\033[35m=========== INFO ===========\033[0m\n");
    printf("FILE: %s", __FILE__);
    printf("APPROACH %d: %s\n", APR, apr_idea);
    printf("  + Time Complexity: O(%s)\n", time_cmplx);
    printf("  + Space Complexity: O(%s)\n", space_cmplx);
    #endif
    
    printf("\033[35m========== TESTCASE ========\033[0m\n");
    struct Case {
        // Input
        int root[20]; int tree_len;
        // Expect
        int exp;

    } cases[] = {
        // {}
        {{1,2,3,4,5,6}, 6, 110},
        {{1,null,2,3,4,null,null,5,6}, 9, 90}
    };

    // Assert test
    struct TreeNode* root;
    int numCases = (int)(sizeof(cases)/sizeof(cases[0]));
    for (int i = 0; i < numCases; ++i) {
        struct Case c = cases[i];

        // Convert array root into TreeNode
        root = tree_create_from_array(c.root, c.tree_len);
        // Solver(...)
        int res = maxProduct(root);
        // Assert Test
        if (assertTest(&res, 0, &c.exp, 0, i, sizeof(int), cmp_int, print_int) == false) {
            // Print input if needed
        }

        free_tree(root);
    }
}

File: scripts/samples/python/treenode_sol.py
from typing import List, Optional, Dict, Set, Tuple
import collections
import heapq
import itertools
import math

# Local include
from inc.test import *
import inc.debug as db
from inc.listnode import *

"""
Problem 1339: Maximum Product of Splitted Binary Tree
LeetCode: https://leetcode.com/problems/maximum-product-of-splitted-binary-tree/
:explain: 
"""

INFO = True
APR = 1
if APR == 1:
    apr_idea = ""
    time_cmplx = ""
    space_cmplx = ""
    
db.DEBUG = True
# Implement
class Solution(object):
    if APR == 1:
        def maxProduct(self, root: Optional[TreeNode]) -> int:
            MOD = 10**9 + 7
            self.max_product = 0

            def _countTotalSum(node: Optional[TreeNode]):
                """DFS count total sum of tree"""
                if not node:
                    return 0
                return node.val + _countTotalSum(node.left) + _countTotalSum(node.right)
            
            total_sum = _countTotalSum(root)

            def _countMaxProduct(node: Optional[TreeNode]):
                """DFS find the max product"""
                if not node:
                    return 0
                left = _countMaxProduct(node.left)
                right = _countMaxProduct(node.right)
                subtree = node.val + left + right
                self.max_product = max(self.max_product, subtree * (total_sum - subtree))
                return subtree
            
            _countMaxProduct(root)
            return self.max_product % MOD
def test1339():
    if INFO:
        print(f"\033[35m=========== INFO ===========\033[0m")
        print(f"FILE: {__file__}")
        print(f"APPROACH {APR}: {apr_idea}")
        print(f"  + Time Complexity: O({time_cmplx})")
        print(f"  + Space Complexity: O({space_cmplx})")

    print("\033[35m========== TESTCASE ========\033[0m")
    cases = [
        ([1,2,3,4,5,6], 110),
        ([1,null,2,3,4,null,null,5,6], 90),
        ([], 0),
        ([0], 0)
    ]

    sol = Solution()

    for i, (root, exp) in enumerate(cases):
        inputs = [root]

        # Convert from list to ListNode/TreeNode
        root = TreeNode.create_from_list(root)
        
        #res = sol.solver(input)
        res = sol.maxProduct(root)
        if assertTest(res, exp, i) == False:
            print(f"   Input    : ", end="")
            for input in inputs:
                print(f"{input}", end=" | ")
            print("\n")

if __name__ == "__main__":
    test1339()